migration_scripts/bulk_load/src/norm_tx_benchmark.clj

```diff
@@ -0,0 +1,215 @@
+(ns norm-tx-benchmark
+  (:require [combinations :refer [&& || |?]]
+            [db]
+            [hex :refer [hex->bytes]]
+            [norm-tx
+             :refer [+transactions+
+                     +tx-calls-pkg+ +tx-calls-mod+ +tx-calls-fun+
+                     +tx-senders+ +tx-recipients+
+                     +tx-input-objects+ +tx-changed-objects+
+                     +tx-digests+ +cp-tx+
+                     +tx-system+]]
+            [alphabase.base58 :as b58]
+            [honey.sql :as sql]))
+
+;; # Normalized Transactions Benchmark
+;;
+;; Testing out similar queries as in `tx-benchmark`, but on the
+;; normalized table.
+;;
+;; The main differences are:
+;;
+;; - Cursors refer to just the transaction sequence number, and not
+;;   also the checkpoint sequence number.
+;;
+;; - Each filter applies to a separate table, the results are joined
+;;   together, and finally the resulting ordered set of transaction
+;;   sequence numbers is used to probe into the `transactions` table.
+;;
+;; - The final join (against `transactions`) is a `LEFT JOIN` to force
+;;   all previous inner joins to be done first. This is to take
+;;   advantage of the fact that the previous joins should limit the
+;;   final result.
+;;
+;; This approach *did not* work as hoped -- in general it yielded
+;; worse performance than the existing schema (see below):
+;;
+;;     [#{:pkg} 0.18803418803418803]
+;;     [#{:mod} 0.1891891891891892]
+;;     [#{:fun} 0.21739130434782608]
+;;     [#{:kind} 0.29770992366412213]
+;;     [#{:cp-<} 0.37988826815642457]
+;;     [#{:changed} 0.38596491228070173]
+;;     [#{:input} 0.42857142857142855]
+;;     [#{:before} 0.45320197044334976]
+;;     [#{:cp->} 0.46875]
+;;     [#{:after} 0.5526315789473685]
+;;     [#{:ids} 0.7857142857142857]
+;;     [#{:cp-=} 1.0]
+;;
+;;       Fig 1. Success rates for queries that include at least one
+;;       filter key.
+;;
+;;
+;;     [#{:changed :pkg} 0.0]
+;;     [#{:mod :changed} 0.0]
+;;     [#{:fun :changed} 0.0]
+;;     [#{:kind :pkg} 0.04878048780487805]
+;;     [#{:mod :kind} 0.05555555555555555]
+;;     [#{:fun :kind} 0.05714285714285714]
+;;     [#{:fun :input} 0.07692307692307693]
+;;     [#{:mod :input} 0.09523809523809523]
+;;
+;;       Fig 2. Worst success rates for queries with combinations of
+;;       at least two filter keys.
+;;
+;; The issue seems to be in Postgres' implementation of Merge Join,
+;; which does not have a way to skip over pages of one relation based
+;; on the next row to merge in another relation (like leapfrog trie
+;; join): Sets of queries with low intersectionality, will result
+;; in (effectively) whole table scans.
```

Comment:
even if merge join was implemented more efficiently, it's also a toss-up for postgres to select the join approach to begin with. And it's not clear whether merge joins are something we always want postgres to leverage 

In more recent tests, I was conducting joins on tx_calls and tx_senders, where filters on the former yielded 7k results while the latter yielded 71 rows total. Due to the much smaller size of the second table, perhaps a hash join would be better? and in cases of significant overlap, perhaps a simple nested loop would be fine

But my takeaway is that if we don't model these relationships and rely purely on joins, we won't get very far in the queries we can handle

Reply:
Indeed -- to test this hypothesis, I needed to disable nested loops and hash joins, to force the merge join. 

> In more recent tests, I was conducting joins on tx_calls and tx_senders, where filters on the former yielded 7k results while the latter yielded 71 rows total. Due to the much smaller size of the second table, perhaps a hash join would be better?

I think postgres would typically perform a hash join in this case, but this still results in essentially a full scan of the high cardinality set, which is what I wanted to avoid. There are variants of merge join that can avoid this (see, e.g. my favourite, [Leapfrog Triejoin](https://arxiv.org/abs/1210.0481)), by using the results from one side of a join to avoid looking at whole regions of another part of the join.

This is hard to pull off in practice because the data needs to be laid out just right, and the join order needs to be chosen carefully to make sure we can skip as much work as possible. I was hoping that some of these techniques had reached production in the last ~10 years, but sadly not (at least in the case of Postgres).

> But my takeaway is that if we don't model these relationships and rely purely on joins, we won't get very far in the queries we can handle

Note that the DBMS can recover some of these relationships for you using analysis, but I agree with this sentiment in general -- using joins for combination queries that span the whole database does not seem like a winning strategy on Postgres.